src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -525,26 +550,77 @@ public TypeSymbolWithAnnotations SetUnknownNullabilityForReferenceTypes()
+
         private sealed class NonLazyType : TypeSymbolWithAnnotations
         {
             private readonly TypeSymbol _typeSymbol;
-            private readonly bool? _isNullable;
+            private readonly bool _nonNullTypes; // PROTOTYPE(NullableReferenceTypes): _nonNullTypes should be lazy to avoid unnecessary cycles.
+            private readonly bool _isAnnotated;
             private readonly ImmutableArray<CustomModifier> _customModifiers;
 
-            public NonLazyType(TypeSymbol typeSymbol, bool? isNullable, ImmutableArray<CustomModifier> customModifiers)
+            // PROTOTYPE(NullableReferenceTypes): [Obsolete("Use explicit NonNullTypes context")]
+            public NonLazyType(TypeSymbol typeSymbol, bool? isNullable, ImmutableArray<CustomModifier> customModifiers) :
+                this(typeSymbol, nonNullTypes: IsNullableToNonNullTypes(isNullable), isAnnotated: IsNullableToIsAnnotated(isNullable), customModifiers)
+            {
+            }
+
+            public NonLazyType(TypeSymbol typeSymbol, bool nonNullTypes, bool isAnnotated, ImmutableArray<CustomModifier> customModifiers)
             {
                 Debug.Assert((object)typeSymbol != null);
                 Debug.Assert(!customModifiers.IsDefault);
-                Debug.Assert(!typeSymbol.IsNullableType() || isNullable == true);
+                Debug.Assert(!typeSymbol.IsNullableType() || isAnnotated);
                 _typeSymbol = typeSymbol;
-                _isNullable = isNullable;
+                _nonNullTypes = nonNullTypes;
+                _isAnnotated = isAnnotated;
                 _customModifiers = customModifiers;
             }
 
-            public sealed override TypeSymbol TypeSymbol => _typeSymbol;
-            public sealed override bool? IsNullable => _isNullable;
+            public override TypeSymbol TypeSymbol => _typeSymbol;
+
+            // PROTOTYPE(NullableReferenceTypes): IsNullable depends on IsValueType which
+            // can lead to cycles when IsNullable is queried early. Replace this property with
+            // the Annotation property that depends on IsAnnotated and NonNullTypes only.
+            public override bool? IsNullable
+            {
+                get
+                {
+                    if (_isAnnotated)
+                    {
+                        return true;
+                    }
+                    if (_nonNullTypes)
+                    {
+                        return false;
+                    }
+                    if (_typeSymbol.IsValueType)
+                    {
+                        return false;
+                    }
+                    return null;
+                }
+            }
+
+            public override bool IsAnnotated => _isAnnotated;
             public override ImmutableArray<CustomModifier> CustomModifiers => _customModifiers;
 
+            protected override AnnotationKind Annotation
```

Comment:
I don't think this property needs to be virtual/override. All implementations should agree based on `IsAnnotated` and `NonNullTypes` properties.
I also am not sure we need this property and enum. The only place that seems to use it (`Equals` method at line 312) could probably just do with comparing `IsAnnotated`).